File: examples/java/linux-node-overview/src/main/java/linuxnode/linux/Disk.java
package linuxnode.linux;

import com.grafana.foundation.cog.Builder;
import com.grafana.foundation.common.FieldTextAlignment;
import com.grafana.foundation.common.TableCellHeight;
import com.grafana.foundation.common.TableFooterOptionsBuilder;
import com.grafana.foundation.dashboard.*;
import com.grafana.foundation.timeseries.PanelBuilder;
import com.grafana.foundation.units.Constants;

import java.util.List;
import java.util.Map;

public class Disk {
    public static PanelBuilder diskIOTimeseries() {
        return Common.defaultTimeSeries().
                title("Disk I/O").
                fillOpacity(0.0).
                unit(Constants.BytesPerSecondSI).
                withTarget(Common.basicPrometheusQuery("rate(node_disk_read_bytes_total{job=\"integrations/raspberrypi-node\", instance=\"$instance\", device!=\"\"}[$__rate_interval])", "{{device}} read")).
                withTarget(Common.basicPrometheusQuery("rate(node_disk_written_bytes_total{job=\"integrations/raspberrypi-node\", instance=\"$instance\", device!=\"\"}[$__rate_interval])", "{{device}} written")).
                withTarget(Common.basicPrometheusQuery("rate(node_disk_io_time_seconds_total{job=\"integrations/raspberrypi-node\", instance=\"$instance\", device!=\"\"}[$__rate_interval])", "{{device}} IO time")).
                overrideByRegexp("/ io time/", List.of(
                        new DynamicConfigValue("unit", Constants.PercentUnit)
                ));
    }

    public static com.grafana.foundation.table.PanelBuilder diskSpaceUsageTable() {
        return new com.grafana.foundation.table.PanelBuilder().
                title("Disk Space Usage").
                align(FieldTextAlignment.AUTO).
                unit(Constants.BytesSI).
                cellHeight(TableCellHeight.SM).
                footer(new TableFooterOptionsBuilder().
                        countRows(false).
                        reducer(List.of("sum"))
                ).
                withTarget(Common.tablePrometheusQuery("max by (mountpoint) (node_filesystem_size_bytes{job=\"integrations/raspberrypi-node\", instance=\"$instance\", fstype!=\"\"})", "A")).
                withTarget(Common.tablePrometheusQuery("max by (mountpoint) (node_filesystem_avail_bytes{job=\"integrations/raspberrypi-node\", instance=\"$instance\", fstype!=\"\"})", "B")).
                withTransformation(transformer1()).
                withTransformation(transformer2()).
                withTransformation(transformer3()).
                withTransformation(transformer4()).
                withTransformation(transformer5()).
                withTransformation(transformer6()).
                withOverride(defaultOverrides("Mounted on", 260)).
                withOverride(defaultOverrides("Size", 93)).
                withOverride(defaultOverrides("Used", 72)).
                withOverride(defaultOverrides("Available", 88)).
                withOverride(complexOverrides());
    }

    private static DataTransformerConfig transformer1() {
        DataTransformerConfig dataTransformerConfig = new DataTransformerConfig();
        dataTransformerConfig.id = "groupBy";
        dataTransformerConfig.options = Map.of(
                "fields", Map.of(
                        "Value #A", Map.of(
                                "aggregations", List.of("lastNotNull"),
                                "operation", "aggregate"
                        ),
                        "Value #B", Map.of(
                                "aggregations", List.of("lastNotNull"),
                                "operation", "aggregate"
                        ),
                        "mountpoint", Map.of(
                                "aggregations", List.of(),
                                "operation", "groupby"
                        )
                )
        );

        return dataTransformerConfig;
    }

    private static DataTransformerConfig transformer2() {
        DataTransformerConfig dataTransformerConfig = new DataTransformerConfig();
        dataTransformerConfig.id = "merge";

        return dataTransformerConfig;
    }

    private static DataTransformerConfig transformer3() {
        DataTransformerConfig dataTransformerConfig = new DataTransformerConfig();
        dataTransformerConfig.id = "calculateField";
        dataTransformerConfig.options = Map.of(
                "alias", "Used",
                "binary", Map.of(
                        "left", "Value #A (lastNotNull)",
                        "operator", "-",
                        "reducer", "sum",
                        "right", "Value #B (lastNotNull)"
                ), "mode", "binary",
                "reduce", Map.of("reducer", "sum"));

        return dataTransformerConfig;
    }

    private static DataTransformerConfig transformer4() {
        DataTransformerConfig dataTransformerConfig = new DataTransformerConfig();
        dataTransformerConfig.id = "calculateField";
        dataTransformerConfig.options = Map.of(
                "alias", "Used, %",
                "binary", Map.of(
                        "left", "Used",
                        "operator", "/",
                        "reducer", "sum",
                        "right", "Value #A (lastNotNull)"
                ),
                "mode", "binary",
                "reduce", Map.of("reducer", "sum"));

        return dataTransformerConfig;
    }

    private static DataTransformerConfig transformer5() {
        DataTransformerConfig dataTransformerConfig = new DataTransformerConfig();
        dataTransformerConfig.id = "organize";
        dataTransformerConfig.options = Map.of(
                "excludeByName", List.of(),
                "indexByName", List.of(),
                "renameByName", Map.of(
                        "Value #A (lastNotNull)", "Size",
                        "Value #B (lastNotNull)", "Available",
                        "mountpoint", "Mounted on"
                )
        );

        return dataTransformerConfig;
    }

    private static DataTransformerConfig transformer6() {
        DataTransformerConfig dataTransformerConfig = new DataTransformerConfig();
        dataTransformerConfig.id = "sortBy";
        dataTransformerConfig.options = Map.of(
                "fields", List.of(),
                "sort", Map.of("field", "Mounted on")
        );

        return dataTransformerConfig;
    }

    private static Builder<DashboardFieldConfigSourceOverrides> defaultOverrides(String options, Integer value) {
        MatcherConfig matcherConfig = new MatcherConfig();
        matcherConfig.id = "byName";
        matcherConfig.options = options;

        DynamicConfigValue dynamicConfigValue = new DynamicConfigValue();
        dynamicConfigValue.id = "custom.width";
        dynamicConfigValue.value = value;
        return new DashboardFieldConfigSourceOverridesBuilder().matcher(matcherConfig).properties(List.of(dynamicConfigValue));
    }

    private static Builder<DashboardFieldConfigSourceOverrides> complexOverrides() {
        MatcherConfig matcherConfig = new MatcherConfig();
        matcherConfig.id = "byName";
        matcherConfig.options = "Used, %";

        DynamicConfigValue dynamicConfigValue1 = new DynamicConfigValue();
        dynamicConfigValue1.id = "unit";
        dynamicConfigValue1.value = "percentunit";

        DynamicConfigValue dynamicConfigValue2 = new DynamicConfigValue();
        dynamicConfigValue2.id = "custom.cellOptions";
        dynamicConfigValue2.value = Map.of("mode", "gradient", "type", "gauge");

        DynamicConfigValue dynamicConfigValue3 = new DynamicConfigValue();
        dynamicConfigValue3.id = "min";
        dynamicConfigValue3.value = 0;

        DynamicConfigValue dynamicConfigValue4 = new DynamicConfigValue();
        dynamicConfigValue4.id = "max";
        dynamicConfigValue4.value = 1;

        return new DashboardFieldConfigSourceOverridesBuilder().
                matcher(matcherConfig).
                properties(List.of(dynamicConfigValue1, dynamicConfigValue2, dynamicConfigValue3, dynamicConfigValue4));
    }
}

File: examples/php/linux-node-overview/src/Linux/Disk.php
<?php

namespace App\Linux;

use Grafana\Foundation\Common as SDKCommon;
use Grafana\Foundation\Dashboard as SDKDashboard;
use Grafana\Foundation\Stat;
use Grafana\Foundation\Table;
use Grafana\Foundation\Timeseries;
use Grafana\Foundation\Units\Constants as Units;

class Disk
{
    public static function rootMountSizeStat(): Stat\PanelBuilder
    {
        return Common::defaultStat()
            ->title('Root mount size')
            ->description('Total capacity on the primary mount point /.')
            ->WithTarget(
                Common::prometheusQuery('node_filesystem_size_bytes{job=~"integrations/(node_exporter|unix)",cluster=~"$cluster",job=~"$job",instance=~"$instance", mountpoint="/", fstype!="rootfs"}', '')
            )
            ->unit(Units::BYTES_IEC)
            ->colorMode(SDKCommon\BigValueColorMode::none())
        ;
    }

    public static function readWriteTimeseries(): Timeseries\PanelBuilder
    {
        return Common::defaultTimeseries()
            ->title('Disk reads/writes')
            ->description('Disk read/writes in bytes per second.')
            ->WithTarget(Common::prometheusQuery('irate(node_disk_read_bytes_total{job=~"integrations/(node_exporter|unix)",cluster=~"$cluster",job=~"$job",instance=~"$instance", device!=""}[$__rate_interval])', '{{ device }} read'))
            ->WithTarget(Common::prometheusQuery('irate(node_disk_written_bytes_total{job=~"integrations/(node_exporter|unix)",cluster=~"$cluster",job=~"$job",instance=~"$instance", device!=""}[$__rate_interval])', '{{ device }} written'))
            ->WithTarget(Common::prometheusQuery('irate(node_disk_io_time_seconds_total{job=~"integrations/(node_exporter|unix)",cluster=~"$cluster",job=~"$job",instance=~"$instance", device!=""}[$__rate_interval])', '{{ device }} io util'))
            ->decimals(0)
            ->unit(Units::BYTES_PER_SECOND_SI)
            ->fillOpacity(1)
            ->gradientMode(SDKCommon\GraphGradientMode::opacity())
            ->overrideByRegexp('/time|used|busy|util/', [
                (new SDKDashboard\DynamicConfigValue(id: 'custom.axisSoftMax', value: 100)),
                (new SDKDashboard\DynamicConfigValue(id: 'custom.drawStyle', value: 'points')),
                (new SDKDashboard\DynamicConfigValue(id: 'unit', value: Units::PERCENT)),
            ])
        ;
    }

    public static function spaceUsageTable(): Table\PanelBuilder
    {
        return (new Table\PanelBuilder())
            ->title('Disk space usage')
            ->description('Disk utilisation in percent, by mountpoint. Some duplication can occur if the same filesystem is mounted in multiple locations.')
            ->footer(
                (new SDKCommon\TableFooterOptionsBuilder())
                    ->countRows(false)
                    ->reducer(['sum'])
            )
            ->datasource(new SDKDashboard\DataSourceRef(uid: '$datasource'))

            ->targets([
                Common::tablePrometheusQuery('node_filesystem_size_bytes{fstype!="", mountpoint!="", job=~"integrations/(node_exporter|unix)",cluster=~"$cluster",job=~"$job",instance=~"$instance"}', 'TOTAL')
                    ->instant(),
                Common::tablePrometheusQuery('node_filesystem_avail_bytes{fstype!="", mountpoint!="", job=~"integrations/(node_exporter|unix)",cluster=~"$cluster",job=~"$job",instance=~"$instance"}', 'FREE')
                    ->legendFormat('{{ mountpoint }} free')
                    ->instant(),
            ])
            ->unit(Units::BYTES_IEC)
            ->thresholds(
                (new SDKDashboard\ThresholdsConfigBuilder())
                    ->mode(SDKDashboard\ThresholdsMode::absolute())
                    ->steps([
                        new SDKDashboard\Threshold(value: null, color: 'light-blue'),
                        new SDKDashboard\Threshold(value: 0.8, color: 'light-yellow'),
                        new SDKDashboard\Threshold(value: 0.9, color: 'light-red'),
                    ])
            )
            // Transformations
            ->withTransformation(
                new SDKDashboard\DataTransformerConfig(
                    id: 'groupBy',
                    options: [
                        "fields" => [
                            "Value #FREE" => [
                                "aggregations" => ["lastNotNull"],
                                "operation" => "aggregate",
                            ],
                            "Value #TOTAL" => [
                                "aggregations" => ["lastNotNull"],
                                "operation" => "aggregate",
                            ],
                            "mountpoint" => [
                                "aggregations" => [],
                                "operation" => "groupby",
                            ],
                        ]
                    ],
                )
            )
            ->withTransformation(new SDKDashboard\DataTransformerConfig(id: 'merge', options: []))
            ->withTransformation(
                new SDKDashboard\DataTransformerConfig(
                    id: 'calculateField',
                    options: [
                        "alias" => "Used",
                        "binary" => [
                            "left" => "Value #TOTAL (lastNotNull)",
                            "operator" => "-",
                            "reducer" => "sum",
                            "right" => "Value #FREE (lastNotNull)",
                        ],
                        "mode" => "binary",
                        "reduce" => ["reducer" => "sum"],
                    ],
                )
            )
            ->withTransformation(
                new SDKDashboard\DataTransformerConfig(
                    id: 'calculateField',
                    options: [
                        "alias" => "Used, %",
                        "binary" => [
                            "left" => "Used",
                            "operator" => "/",
                            "reducer" => "sum",
                            "right" => "Value #TOTAL (lastNotNull)",
                        ],
                        "mode" => "binary",
                        "reduce" => ["reducer" => "sum"],
                    ],
                )
            )
            ->withTransformation(
                new SDKDashboard\DataTransformerConfig(
                    id: 'organize',
                    options: [
                        "excludeByName" => [],
                        "indexByName" => [
                            "Used" => 3,
                            "Used, %" => 4,
                            "Value #FREE (lastNotNull)" => 2,
                            "Value #TOTAL (lastNotNull)" => 1,
                            "mountpoint" => 0,
                        ],
                        "renameByName" => [
                            "Value #TOTAL (lastNotNull)" => "Size",
                            "Value #FREE (lastNotNull)" => "Available",
                            "mountpoint" => "Mounted on",
                        ],
                    ],
                )
            )
            ->withTransformation(new SDKDashboard\DataTransformerConfig(
                id: 'sortBy',
                options: [
                    "fields" => [],
                    "sort" => [["field" => "Mounted on", "desc" => false]],
                ],
            ))
            // Overrides
            ->overrideByName('Mounted on', [
                new SDKDashboard\DynamicConfigValue(id: 'custom.width', value: 260),
            ])
            ->overrideByName('Size', [
                new SDKDashboard\DynamicConfigValue(id: 'custom.width', value: 80),
            ])
            ->overrideByName('Used', [
                new SDKDashboard\DynamicConfigValue(id: 'custom.width', value: 80),
            ])
            ->overrideByName('Available', [
                new SDKDashboard\DynamicConfigValue(id: 'custom.width', value: 80),
            ])
            ->overrideByName('Used, %', [
                new SDKDashboard\DynamicConfigValue(id: 'custom.cellOptions', value: [
                    'mode' => 'basic',
                    'type' => 'gauge',
                ]),
                new SDKDashboard\DynamicConfigValue(id: 'min', value: 0),
                new SDKDashboard\DynamicConfigValue(id: 'max', value: 1),
                new SDKDashboard\DynamicConfigValue(id: 'unit', value: Units::PERCENT_UNIT),
            ])
        ;
    }
}
